Update developer guide to support new patch version of Spark (#109)
dotnet/spark: docs/developer-guide.md

> # Developer Guide

Reviewer: Spark .NET Developer Guide? (since we want to differentiate from an Application Developer Guide)

## Table of Contents

> - [How to Debug](#how-to-debug)

Reviewer: We'd have to provide both local and remote debugging (we can start with local but we should distinguish).

- [How to Support New Spark Releases](#how-to-support-new-spark-releases)

## How to debug

> ### Debugging .NET application

Reviewer: ```suggestion
### Debugging Spark .NET application
```

  <path-to-microsoft-spark-jar> \
  <path-to-your-app-exe> <argument(s)-to-your-app>
```

## How to Support New Spark Releases

> 

Reviewer: ```suggestion

We encourage developers to first read Apache Spark's [Versioning Policy](https://spark.apache.org/versioning-policy.html) and [Semantic Versioning](https://semver.org/) to gain the most out of the instructions below. 

At a high-level, Spark's versions are: **[MAJOR].[FEATURE].[MAINTENANCE]**. We will cover the upgrade path for each type of version separately below (in increasing order of effort required).
```

> ### New patch version release

Reviewer: ```suggestion
### [MAINTENANCE]: Upgrading for a Patch Release Version
```

> Apache Spark's new patch release involves only the internal changes such as bug fixes, etc. Thus, it is straightfoward to support a new patch version release.

Reviewer: ```suggestion
Since Apache Spark's [MAINTENANCE] releases involve only internal changes (e.g., bug fixes etc.), it is straightforward to upgrade the code base to support a [MAINTENANCE] release. The steps to do this are below:
```

1. In the corresponding `pom.xml`, update the `spark.version` value to the newly released version.

>    * For example, if a new patch release is 2.4.3, you will update `src/scala/microsoft-spark-2.4.x/pom.xml` to have `<spark.version>2.4.3</spark.version>`.

Reviewer: Can you link to pom.xml?

2. Update `DotnetRunner.supportedSparkVersions` to include the newly released version.
   * For example, if a new patch release is 2.4.3, you will update `src/scala/microsoft-spark-2.4.x/src/main/scala/org/apache/spark/deploy/DotnetRunner.scala`.

> 3. Update the build pipeline to include E2E testing for the newly released version.

Reviewer: Can we provide more information here? For instance, link to the yml file and explain what lines to change?

Author: I created a link to yaml and right below you have a link to the commit for an example.

> Refer to [] for an example.

Author: This will be updated to include commit version when #108 is checked in.

> ### New minor version release

Reviewer: ```suggestion
### [FEATURE]: Upgrading for a Minor Release Version
```

*WIP*

> ### New major version release

Reviewer: ```suggestion
### [MAJOR]: Upgrading for a Major Release Version
```
